source/common/common/inline_map.h

```diff
@@ -0,0 +1,495 @@
+
+    if (auto it = inline_keys_map_.find(key); it == inline_keys_map_.end()) {
+      return absl::nullopt;
+    } else {
+      return it->second;
+    }
+  }
+
+  /**
+   * Finalize this descriptor. No further changes are allowed after this point. This guaranteed that
+   * all map created by the process have the same variable size and custom inline key adding. This
+   * method should only be called once.
+   */
+  void finalize() {
+    ASSERT(!finalized_, "Cannot finalize() an already finalized descriptor");
+    finalized_ = true;
+  }
+
+  /**
+   * @return true if the descriptor is finalized.
+   */
+  bool finalized() const { return finalized_; }
+
+  /**
+   * @return the debug string of this descriptor. This will return all inline keys that joined by
+   * the given separator.
+   */
+  std::string debugString(absl::string_view separator = ",") const {
+    ASSERT(finalized_, "Cannot fetch debug string before finalize()");
+    return absl::StrJoin(inline_keys_, separator);
+  }
+
+private:
+  template <class Key, class Value> friend class InlineMap;
+
+  /**
+   * Get the inline keys map that contains all inline keys and their handle. May only be called
+   * after finalize().
+   */
+  const InlineKeysMap& inlineKeysMap() const {
+    ASSERT(finalized_, "Cannot fetch registration map before finalize()");
+    return inline_keys_map_;
+  }
+
+  /**
+   * Get the array that contains all added inline keys. May only be called after finalize().
+   */
+  const InlineKeys& inlineKeys() const {
+    ASSERT(finalized_, "Cannot fetch registration set before finalize()");
+    return inline_keys_;
+  }
+
+  /**
+   * Get the number of the inline keys in this descriptor. May only be called after finalize().
+   */
+  uint64_t inlineKeysNum() const {
+    ASSERT(finalized_, "Cannot fetch registration map before finalize()");
+    return inline_keys_map_.size();
+  }
+
+  bool finalized_{};
+  InlineKeys inline_keys_;
+  InlineKeysMap inline_keys_map_;
+};
+
+/**
+ * This is the inline map that could be used as an alternative of normal hash map to store the
+ * key/value pairs.
+ */
+template <class Key, class Value> class InlineMap : public InlineStorage {
+private:
+  using TypedInlineMapDescriptor = InlineMapDescriptor<Key>;
+  using DynamicHashMap = absl::flat_hash_map<Key, Value>;
+
+public:
+  using Handle = typename TypedInlineMapDescriptor::Handle;
+  using ValueRef = std::reference_wrapper<Value>;
+
+  /**
+   * Get the entry by the given key. Heterogeneous lookup is supported here.
+   * @param key the key to lookup.
+   * @return the entry if the key is found, otherwise return null reference.
+   */
+  template <class GetKey> OptRef<const Value> get(const GetKey& key) const {
+    // Because the dynamic key is used here, try the normal map entry first.
+    if (auto it = dynamic_entries_.find(key); it != dynamic_entries_.end()) {
+      return it->second;
+    }
+
+    if (auto entry_id = inlineLookup(key); entry_id.has_value()) {
```

Comment:
explicit type as the `absl::optional<uint64_t>` is not obvious from context, here and below.